File: CLDemo-OC/CLDemo-OC/Class/CLRecordeEncodeController/CLRecorder/CLMp3Encoder.m
//
//  CLMp3Encoder.m
//  CLDemo
//
//  Created by JmoVxia on 2020/6/24.
//  Copyright © 2020 JmoVxia. All rights reserved.
//

#import "CLMp3Encoder.h"
#import "lame.h"

const int MP3_BUFF_SIZE = 4096;

@interface CLMp3Encoder ()
{
    lame_t lame;
    unsigned char mp3_buffer[MP3_BUFF_SIZE];
    short int *pcmData1;
    short int *pcmData2;
    dispatch_queue_t _encodeQueue;
}

@end

@implementation CLMp3Encoder

- (instancetype)init {
    self = [super init];
    if (self) {
        _encodeQueue = dispatch_queue_create("CLMp3Encoder.encodeQueue", DISPATCH_QUEUE_SERIAL);
        _inputSampleRate = [AVAudioSession sharedInstance].sampleRate;
        _outputSampleRate = [AVAudioSession sharedInstance].sampleRate;
        _outputChannelsPerFrame = 1;
        _bitRate = 16;
        _quality = 5;
    }
    return self;
}
- (void)run {
    if (lame == NULL) {
        lame = lame_init();
        lame_set_in_samplerate(lame,_inputSampleRate);
        lame_set_quality (lame, _quality);
        lame_set_VBR(lame, vbr_default);
        lame_set_brate(lame, _bitRate);
        lame_set_out_samplerate(lame, _outputSampleRate);
        lame_set_num_channels(lame, _outputChannelsPerFrame);
        lame_init_params(lame);
    }
    _isRunning = YES;
}
- (void)stop {
    if (lame) {
        lame_close(lame);
        lame = NULL;
    }
    _isRunning = NO;
}
- (void)processAudioBufferList: (AudioBufferList)audioBufferList {
    if(!_isRunning){
        return;
    }
    dispatch_async(_encodeQueue, ^{
        int channels = audioBufferList.mBuffers[0].mNumberChannels;
        short int *pcmData = audioBufferList.mBuffers[0].mData;
        int size = audioBufferList.mBuffers[0].mDataByteSize / (sizeof(short int) * channels);
        int bytesWritten = 0;
        NSData *data = nil;
        if (channels == 2) {
            bytesWritten = lame_encode_buffer_interleaved(self->lame, pcmData, size, self->mp3_buffer, MP3_BUFF_SIZE);
        }else if (channels == 1) {
            if (self->pcmData1 == NULL) {
                self->pcmData1 = pcmData;
            }
            if (self->pcmData2 == NULL) {
                self->pcmData2 = pcmData;
            }
            if (self->pcmData1 && self->pcmData2) {
                bytesWritten = lame_encode_buffer(self->lame, self->pcmData1, self->pcmData2,size, self->mp3_buffer, MP3_BUFF_SIZE);
                self->pcmData1 = NULL;
                self->pcmData2 = NULL;
            }
        }
        if (bytesWritten && self.processingEncodedData) {
            data = [[NSData alloc] initWithBytes:self->mp3_buffer length:bytesWritten];
            self.processingEncodedData(data);
        }
    });
}
- (void)dealloc {
    [self stop];
}
@end

File: CLDemo-Swift/CLDemo-Swift/CLRecorder/CLMp3Encoder.swift
//
//  CLMp3Encoder.swift
//  CKD
//
//  Created by Chen JmoVxia on 2024/5/28.
//  Copyright © 2024 JmoVxia. All rights reserved.
//

import AVFAudio
import Foundation

class CLMp3Encoder {
    deinit {
        stop()
    }

    private static let mp3BufferSize = 4096

    private let encodeQueue = DispatchQueue(label: "CKDMp3Encoder.encodeQueue")

    private var lame: OpaquePointer?
    private var mp3Buffer = [UInt8](repeating: 0, count: CLMp3Encoder.mp3BufferSize)
    private var pcmData1: UnsafeMutablePointer<Int16>?
    private var pcmData2: UnsafeMutablePointer<Int16>?
    private(set) var isRunning = false

    var inputSampleRate = AVAudioSession.sharedInstance().sampleRate
    var outputSampleRate = AVAudioSession.sharedInstance().sampleRate
    var outputChannelsPerFrame = 1
    var bitRate: Int32 = 16
    var quality: Int32 = 5
    var processingEncodedData: ((Data) -> Void)?
}

extension CLMp3Encoder {
    func run() {
        guard lame == nil else { return }

        lame = lame_init()
        guard let lame else { return }

        lame_set_in_samplerate(lame, Int32(inputSampleRate))
        lame_set_quality(lame, quality)
        lame_set_VBR(lame, vbr_default)
        lame_set_brate(lame, bitRate)
        lame_set_out_samplerate(lame, Int32(outputSampleRate))
        lame_set_num_channels(lame, Int32(outputChannelsPerFrame))
        lame_init_params(lame)

        isRunning = true
    }

    func stop() {
        guard let lame else { return }
        lame_close(lame)
        self.lame = nil
        isRunning = false
    }

    func processAudioBufferList(_ audioBufferList: AudioBufferList) {
        guard isRunning else { return }
        encodeQueue.async { [weak self] in
            guard let self else { return }
            guard let audioData = audioBufferList.mBuffers.mData?.assumingMemoryBound(to: Int16.self) else { return }

            let channels = Int(audioBufferList.mBuffers.mNumberChannels)
            let dataSize = Int(audioBufferList.mBuffers.mDataByteSize) / (MemoryLayout<Int16>.size * channels)
            var bytesWritten: Int32 = 0

            if channels == 2 {
                bytesWritten = lame_encode_buffer_interleaved(lame, audioData, Int32(dataSize), &mp3Buffer, Int32(CLMp3Encoder.mp3BufferSize))
            } else if channels == 1 {
                if pcmData1 == nil { pcmData1 = audioData }
                if pcmData2 == nil { pcmData2 = audioData }

                if let pcmData1, let pcmData2 {
                    bytesWritten = lame_encode_buffer(lame, pcmData1, pcmData2, Int32(dataSize), &mp3Buffer, Int32(CLMp3Encoder.mp3BufferSize))
                    self.pcmData1 = nil
                    self.pcmData2 = nil
                }
            }
            guard bytesWritten > 0 else { return }
            processingEncodedData?(Data(bytes: mp3Buffer, count: Int(bytesWritten)))
        }
    }
}
